File: GCJ/2019Q/P3.cpp
#include<bits/stdc++.h>
using namespace std;

#define pb push_back
#define ll long long
#define maxn 300005
#define fr(i,j,k) for(int i=j;i<k;i++)
#define f(n) fr(i,0,n)
#define f1(n) fr(i,1,n+1)
#define ms(i) memset(i,-1,sizeof(i));

int main(){
	int t;
	cin >> t;
	for(int c = 1 ; c <=t;c++){

		int n,k;
		cin >> k >> n;
		n++;
		int cont[n+5];
		int num[n+5];
		f(n-1){
			cin >>cont[i];
		}
		for(int i=0;i<n;i++){
			if(cont[i]!=cont[i+1]){
				num[i+1]=__gcd(cont[i],cont[i+1]);
				//cout<<num[i+1]<<endl;
				for(int j=i;j>=0;j--){
					num[j]=cont[j]/num[j+1];
					//cout<<num[j]<<endl;
				}
				for(int j=i+2;j<n;j++){
					num[j]=cont[j-1]/num[j-1];
					//cout<<num[j]<<endl;
				}
				break;
			}
			
		}
		set<int>st;
		f(n){
			st.insert(num[i]);
		}
		map<int,char>mp;
		char now='A';
		for(auto i:st){
			mp[i]=now;
			now++;
		}
		printf("Case #%d: ",c);
		for(int i=0;i<n;i++){
			cout<<mp[num[i]];
		}
		cout<<endl;
	}
}

File: GCJ/2019Q/P3.py
import math
t=int(input())
for c in range(1,t+1):
    R=lambda:map(int,input().split())
    k,n=R()
    n+=1
    cont=[]
    num=[]
    tmp=[]
    cont = [int(i) for i in input().split()]
    for _ in range (0,n+5):
        num.append(0)
    for i in range (0,n):
        if cont[i]!=cont[i+1]:
            num[i+1]=math.gcd(cont[i],cont[i+1])
            for j in range (i+2,n):
                num[j]=cont[j-1]//num[j-1]
            for j in range (i,-1,-1):
                num[j]=cont[j]//num[j+1]
            break
                
		
    for i in range(0,n):
        tmp.append(num[i])
    tmp.sort()
    now = ord('A')
    d={tmp[0]:chr(now)}
    print("Case #",end="")
    print(c,end="")
    print(": ",end="")
    for i in range(1,n):
        if tmp[i]!=tmp[i-1]:
            now = now +1
            d[tmp[i]]=chr(now)
    for i in range(0,n):
        print(d[num[i]],end="")
    print()

